File: packages/agent-server/src/lib/wechat-db.ts
/**
 * WeChat SQLCipher database query helpers.
 *
 * Uses the `sqlcipher` CLI (built from source in the Docker image)
 * to query WeChat's encrypted databases.
 */

import { execSync } from "child_process";
import fs from "fs";
import path from "path";

/**
 * Query a WeChat SQLCipher database and return parsed rows.
 *
 * Uses `-json` output mode for structured results. Falls back to raw
 * string output if json mode produces no results.
 */
export function queryWechatDb(
  dbPath: string,
  hexKey: string,
  sql: string
): Record<string, unknown>[] {
  const input = [
    `PRAGMA key = "x'${hexKey}'";`,
    `PRAGMA cipher_compatibility = 4;`,
    `.mode json`,
    sql,
  ].join("\n");

  const result = execSync(`sqlcipher "${dbPath}"`, {
    input,
    timeout: 10_000,
    encoding: "utf-8",
    stdio: ["pipe", "pipe", "pipe"],
  });

  const trimmed = result.trim();
  if (!trimmed) return [];

  // sqlcipher outputs "ok" for PRAGMAs before the JSON array.
  // Find the JSON array in the output.
  const jsonStart = trimmed.indexOf("[");
  if (jsonStart === -1) return [];

  const jsonStr = trimmed.slice(jsonStart);
  if (jsonStr === "[]") return [];

  try {
    return JSON.parse(jsonStr);
  } catch {
    return [];
  }
}

/**
 * Find the WeChat process PID.
 *
 * pgrep may return multiple PIDs (parent su wrapper + actual wechat).
 * We need the one that actually has open db_storage files, so we
 * try each candidate and return the first with readable /proc/pid/fd.
 */
export function findWechatPid(): number | null {
  try {
    const output = execSync("pgrep -f /usr/bin/wechat", {
      encoding: "utf-8",
      timeout: 5_000,
      stdio: ["pipe", "pipe", "pipe"],
    }).trim();

    const pids = output.split("\n").map(s => parseInt(s, 10)).filter(n => !isNaN(n) && n > 0);

    // Return the PID with the most open file descriptors (the real process)
    let bestPid: number | null = null;
    let bestFdCount = 0;
    for (const pid of pids) {
      try {
        const fdCount = fs.readdirSync(`/proc/${pid}/fd`).length;
        if (fdCount > bestFdCount) {
          bestFdCount = fdCount;
          bestPid = pid;
        }
      } catch {
        // Can't read this PID's fds
      }
    }
    return bestPid;
  } catch {
    // pgrep returns exit code 1 if no match
  }
  return null;
}

/**
 * Detect the WeChat account directory by scanning /proc/<pid>/fd
 * for open database files.
 *
 * Returns the account directory name (e.g. "wxid_abc123_def456")
 * or null if not found.
 */
export function findAccountDir(wechatPid: number): string | null {
  try {
    const fdDir = `/proc/${wechatPid}/fd`;
    const fds = fs.readdirSync(fdDir);

    for (const fd of fds) {
      try {
        const target = fs.readlinkSync(path.join(fdDir, fd));
        if (target.includes("db_storage") && target.endsWith(".db")) {
          const idx = target.indexOf("xwechat_files/");
          if (idx >= 0) {
            const rest = target.slice(idx + "xwechat_files/".length);
            const accountDir = rest.split("/")[0];
            if (accountDir) return accountDir;
          }
        }
      } catch {
        // Skip unreadable fds
        continue;
      }
    }
  } catch {
    // /proc not available or permission denied
  }
  return null;
}

/**
 * List all .db files that exist on disk for a given account.
 * Returns db names like ["session.db", "contact.db", ...].
 */
export function listAccountDbs(accountDir: string): string[] {
  const dbNames: string[] = [];

  for (const base of [
    path.join("/home/wechat/xwechat_files", accountDir),
    path.join("/home/wechat/Documents/xwechat_files", accountDir),
  ]) {
    const dbStorageDir = path.join(base, "db_storage");
    if (!fs.existsSync(dbStorageDir)) continue;

    // Scan subdirectories for .db files
    for (const subDir of fs.readdirSync(dbStorageDir)) {
      const subPath = path.join(dbStorageDir, subDir);
      if (!fs.statSync(subPath).isDirectory()) continue;

      for (const file of fs.readdirSync(subPath)) {
        if (file.endsWith(".db")) {
          dbNames.push(file);
        }
      }
    }
    if (dbNames.length > 0) return dbNames;
  }

  return dbNames;
}

/**
 * Get the full path to a WeChat database file.
 *
 * Maps db names like "session.db" to their subdirectory:
 *   session.db → db_storage/session/session.db
 *   contact.db → db_storage/contact/contact.db
 *   message_0.db → db_storage/message/message_0.db
 */
export function getDbPath(accountDir: string, dbName: string): string {
  // Map db names to their subdirectories
  const subDirMap: Record<string, string> = {
    "contact.db": "contact",
    "contact_fts.db": "contact",
    "session.db": "session",
    "message_0.db": "message",
    "message_fts.db": "message",
    "message_resource.db": "message",
    "biz_message_0.db": "message",
    "media_0.db": "message",
    "general.db": "general",
    "hardlink.db": "hardlink",
    "head_image.db": "head_image",
    "emoticon.db": "emoticon",
    "favorite.db": "favorite",
    "favorite_fts.db": "favorite",
    "sns.db": "sns",
    "bizchat.db": "bizchat",
  };

  const subDir = subDirMap[dbName] || dbName.replace(/\.db$/, "");

  // Check both possible base paths
  for (const base of [
    path.join("/home/wechat/xwechat_files", accountDir),
    path.join("/home/wechat/Documents/xwechat_files", accountDir),
  ]) {
    const fullPath = path.join(base, "db_storage", subDir, dbName);
    if (fs.existsSync(fullPath)) return fullPath;
  }

  // Default to first path even if not found
  return path.join(
    "/home/wechat/xwechat_files",
    accountDir,
    "db_storage",
    subDir,
    dbName
  );
}

File: packages/agent-server-rust/src/tools/wechat_db.rs
use std::path::{Path, PathBuf};
use std::process::Command;

/// Query a WeChat database and return parsed rows.
/// Uses sqlcipher CLI with `-json` output mode.
pub fn query_wechat_db(
    db_path: &str,
    hex_key: &str,
    sql: &str,
) -> Vec<serde_json::Value> {
    let input = format!(
        "PRAGMA key = \"x'{hex_key}'\";\nPRAGMA cipher_compatibility = 4;\n.mode json\n{sql}"
    );

    let output = Command::new("sqlcipher")
        .arg(db_path)
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
        .and_then(|mut child| {
            use std::io::Write;
            if let Some(stdin) = child.stdin.as_mut() {
                let _ = stdin.write_all(input.as_bytes());
            }
            child.wait_with_output()
        });

    let output = match output {
        Ok(o) => o,
        Err(_) => return Vec::new(),
    };

    let stdout = String::from_utf8_lossy(&output.stdout);
    let trimmed = stdout.trim();
    if trimmed.is_empty() {
        return Vec::new();
    }

    // Find the JSON array in the output (PRAGMAs output "ok" first)
    let json_start = match trimmed.find('[') {
        Some(i) => i,
        None => return Vec::new(),
    };

    let json_str = &trimmed[json_start..];
    serde_json::from_str(json_str).unwrap_or_default()
}

/// Find the WeChat process PID.
pub fn find_wechat_pid() -> Option<i64> {
    let output = Command::new("pgrep")
        .args(["-f", "/usr/bin/wechat"])
        .output()
        .ok()?;

    let stdout = String::from_utf8_lossy(&output.stdout);
    let pids: Vec<i64> = stdout
        .split_whitespace()
        .filter_map(|s| s.parse().ok())
        .collect();

    // Return the PID with the most open file descriptors
    let mut best_pid: Option<i64> = None;
    let mut best_fd_count = 0;

    for pid in pids {
        let fd_dir = format!("/proc/{pid}/fd");
        if let Ok(entries) = std::fs::read_dir(&fd_dir) {
            let count = entries.count();
            if count > best_fd_count {
                best_fd_count = count;
                best_pid = Some(pid);
            }
        }
    }

    best_pid
}

/// Detect the WeChat account directory by scanning /proc/<pid>/fd.
pub fn find_account_dir(wechat_pid: i64) -> Option<String> {
    let fd_dir = format!("/proc/{wechat_pid}/fd");
    let entries = std::fs::read_dir(&fd_dir).ok()?;

    for entry in entries.flatten() {
        if let Ok(target) = std::fs::read_link(entry.path()) {
            let target_str = target.to_string_lossy();
            if target_str.contains("db_storage") && target_str.ends_with(".db") {
                if let Some(idx) = target_str.find("xwechat_files/") {
                    let rest = &target_str[idx + "xwechat_files/".len()..];
                    if let Some(account_dir) = rest.split('/').next() {
                        if !account_dir.is_empty() {
                            return Some(account_dir.to_string());
                        }
                    }
                }
            }
        }
    }

    None
}

/// List all .db files that exist on disk for a given account.
pub fn list_account_dbs(account_dir: &str) -> Vec<String> {
    let base_paths = [
        format!("/home/wechat/xwechat_files/{account_dir}"),
        format!("/home/wechat/Documents/xwechat_files/{account_dir}"),
    ];

    for base in &base_paths {
        let db_storage = PathBuf::from(base).join("db_storage");
        if !db_storage.exists() {
            continue;
        }

        let mut db_names = Vec::new();
        if let Ok(sub_dirs) = std::fs::read_dir(&db_storage) {
            for sub_dir in sub_dirs.flatten() {
                if sub_dir.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                    if let Ok(files) = std::fs::read_dir(sub_dir.path()) {
                        for file in files.flatten() {
                            let name = file.file_name().to_string_lossy().to_string();
                            if name.ends_with(".db") {
                                db_names.push(name);
                            }
                        }
                    }
                }
            }
        }

        if !db_names.is_empty() {
            return db_names;
        }
    }

    Vec::new()
}

/// Get the full path to a WeChat database file.
pub fn get_db_path(account_dir: &str, db_name: &str) -> String {
    let sub_dir_map: &[(&str, &str)] = &[
        ("contact.db", "contact"),
        ("contact_fts.db", "contact"),
        ("session.db", "session"),
        ("message_0.db", "message"),
        ("message_fts.db", "message"),
        ("message_resource.db", "message"),
        ("biz_message_0.db", "message"),
        ("media_0.db", "message"),
        ("general.db", "general"),
        ("hardlink.db", "hardlink"),
        ("head_image.db", "head_image"),
        ("emoticon.db", "emoticon"),
        ("favorite.db", "favorite"),
        ("favorite_fts.db", "favorite"),
        ("sns.db", "sns"),
        ("bizchat.db", "bizchat"),
    ];

    let sub_dir = sub_dir_map
        .iter()
        .find(|(name, _)| *name == db_name)
        .map(|(_, dir)| *dir)
        .unwrap_or_else(|| db_name.strip_suffix(".db").unwrap_or(db_name));

    let base_paths = [
        format!("/home/wechat/xwechat_files/{account_dir}"),
        format!("/home/wechat/Documents/xwechat_files/{account_dir}"),
    ];

    for base in &base_paths {
        let full_path = Path::new(base)
            .join("db_storage")
            .join(sub_dir)
            .join(db_name);
        if full_path.exists() {
            return full_path.to_string_lossy().to_string();
        }
    }

    // Default to first path
    Path::new(&base_paths[0])
        .join("db_storage")
        .join(sub_dir)
        .join(db_name)
        .to_string_lossy()
        .to_string()
}
